docs/SPIRVVersionsAndExtensionsHandling.rst

```diff
@@ -0,0 +1,247 @@
+corresponding to a spec version which is not included into set of allowed
+SPIR-V versions, translator emits error.
+
+SPIR-V Extensions
+=================
+
+SPIR-V Generation step
+----------------------
+
+By default, translator doesn't use any extensions. If it required to enable
+certain extension, the following command line option can be used:
+
+* ``--spirv-ext=`` - allows to control list of allowed/disallowed extensions.
+
+Valid value for this option is comma-separated list of extension names prefixed
+with ``+`` or ``-`` - plus means allow to use extension, minus means disallow
+to use extension. There is one more special value which can be used as extension
+name in this option: ``all`` - it affects all extension which are known to the
+translator.
+
+If ``--spirv-ext`` contains name of extension which is not know for the
+translator, it will emit error.
+
+Examples:
+
+* ``--spirv-ext=+SPV_KHR_no_integer_wrap_decoration,+SPV_INTEL_subgroups``
+* ``--spirv-ext=+all,-SPV_INTEL_fpga_loop_controls``
+
+.. warning::
+   Extension name cannot be allowed and disallowed at the same time: for inputs
+   like ``--spirv-ext=+SPV_INTEL_subgroups,-SPV_INTEL_subgroups`` translator
+   will emit error about invalid arguments.
+
+.. note::
+   Since by default during SPIR-V generation all extensions are disabled, this
+   means that ``-all,`` is implicitly added at the beggining of the
+   ``-spirv-ext`` value.
+
+If the translator encounters something that cannot be represented by set of
+allowed SPIR-V extensions (which might be empty), it does one of the following
+things:
+
+* ignores LLVM IR entity in the input file.
+
+  For example, ``nsw``/``nuw`` LLVM IR attributes can be ignored if it is not
+  allowed to generate SPIR-V 1.4 and ``SPV_KHR_no_integer_wrap_decoration``
+  extension is disallowed.
+
+* tries to represent LLVM IR entity with allowed instructions.
+
+  Translator could translate calls to a new built-in functions defined by some
+  extensions as usual call instructions without using special SPIR-V
+  instructions.
+
+  However, this could result in a strange SPIR-V and most likely will lead to
+  errors during consumption. Having that, translator should emit errors if it
+  encounters a call to a built-in function from an extension which must be
+  represented as a special SPIR-V instruction from extension which wasn't
+  allowed to be used. I.e. if translator knows that this certain LLVM IR entity
+  belongs to an extension functionality and this extension is disallowed, it
+  should emit error rather than emulating it.
+
+  This way of emulating missing instructions still might be taken by the
+  translator if it encounters something that can be represented in SPIR-V more
+  easily in presence of some extension, but can be legally emulated without it.
+  I.e. in situations similar to ``OpPtrEqual`` in SPIR-V less than 1.4
+
+  **TODO:** add example if such exists.
```

Comment:
Do you need help verifying this?

Reply:
@StuartDBrady, if you have some examples or know how to easily get such, please let me know, it would be great

Reply:
It turned out that it is not so easy to find some extension functionality which can be easily emulated with core functionality (what is the point of the extension then?)

The only one thing that I've found so far is  that `OpSubgroupAllKHR` instruction from [SPV_KHR_subgroup_vote](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_subgroup_vote.html) can be replaced with `OpGroupAll` from the core SPIR-V spec. But, I haven't verified yet that `OpGroupAll` is available in the older SPIR-V versions, i.e. it is legal to replace one for another or not

Reply:
So, I haven't found any good examples, so, I will remove this TODO. If we find something good to place here we can always update documentation later